Source/Core/VideoCommon/LightingShaderGen.cpp

```diff
@@ -83,94 +104,135 @@ static void GenerateLightShader(ShaderCode& object, const LightingUidData& uid_d
+      const auto attnfunc = static_cast<AttenuationFunc>(attn >> 1);
+      const auto diffusefunc = static_cast<DiffuseFunc>((uid_data.diffusefunc >> (2 * chan)) & 3);
+      const u32 light_mask =
+          (uid_data.light_mask >> (NUM_XF_LIGHTS * chan)) & ((1 << NUM_XF_LIGHTS) - 1);
+      object.Write("  // Color attenuation function: {}\n", attnfunc);
+      object.Write("  // Color diffuse function: {}\n", diffusefunc);
+      object.Write("  // Color light mask: {:08b}\n", light_mask);
+      for (u32 light = 0; light < NUM_XF_LIGHTS; light++)
       {
-        if ((uid_data.light_mask & (1 << (i + 8 * j))) != 0)
-          GenerateLightShader(object, uid_data, i, j, false);
+        if ((light_mask & (1 << light)) != 0)
+        {
+          GenerateLightShader(object, uid_data, light, enable_attn, attnfunc, diffusefunc, false);
+        }
       }
     }
-    if ((uid_data.enablelighting & (1 << (j + 2))) != 0)  // Alpha lights
+    if (alpha_enable)
     {
-      for (int i = 0; i < 8; ++i)
+      const u32 attn = (uid_data.attnfunc >> (2 * chan_a)) & 3;
+      const auto enable_attn = static_cast<bool>(attn & 1);
+      const auto attnfunc = static_cast<AttenuationFunc>(attn >> 1);
+      const auto diffusefunc = static_cast<DiffuseFunc>((uid_data.diffusefunc >> (2 * chan_a)) & 3);
+      const u32 light_mask =
+          (uid_data.light_mask >> (NUM_XF_LIGHTS * chan_a)) & ((1 << NUM_XF_LIGHTS) - 1);
+      object.Write("  // Alpha attenuation function: {}\n", attnfunc);
+      object.Write("  // Alpha diffuse function: {}\n", diffusefunc);
+      object.Write("  // Alpha light mask function: {:08b}\n", light_mask);
+      for (u32 light = 0; light < NUM_XF_LIGHTS; light++)
       {
-        if ((uid_data.light_mask & (1 << (i + 8 * (j + 2)))) != 0)
-          GenerateLightShader(object, uid_data, i, j + 2, true);
+        if ((light_mask & (1 << light)) != 0)
+        {
+          GenerateLightShader(object, uid_data, light, enable_attn, attnfunc, diffusefunc, true);
+        }
       }
     }
-    object.Write("lacc = clamp(lacc, 0, 255);\n");
-    object.Write("{}{} = float4((mat * (lacc + (lacc >> 7))) >> 8) / 255.0;\n", dest, j);
+    object.Write("  lacc = clamp(lacc, 0, 255);\n");
+    object.Write("  {}{} = float4((mat * (lacc + (lacc >> 7))) >> 8) / 255.0;\n", dest, chan);
     object.Write("}}\n");
   }
 }
 
 void GetLightingShaderUid(LightingUidData& uid_data)
 {
-  for (u32 j = 0; j < NUM_XF_COLOR_CHANNELS; j++)
+  for (u32 chan = 0; chan < NUM_XF_COLOR_CHANNELS; chan++)
   {
-    uid_data.matsource |= static_cast<u32>(xfmem.color[j].matsource.Value()) << j;
-    uid_data.matsource |= static_cast<u32>(xfmem.alpha[j].matsource.Value()) << (j + 2);
-    uid_data.enablelighting |= xfmem.color[j].enablelighting << j;
-    uid_data.enablelighting |= xfmem.alpha[j].enablelighting << (j + 2);
+    const u32 chan_a = chan + NUM_XF_COLOR_CHANNELS;  // Data for alpha is stored after all colors
```

Comment:
It's likely nicer from a readability standpoint to put the comment above the line rather than off to the side in this instance